igvc/src/cnn/cnn.py

```diff
@@ -0,0 +1,173 @@
+        # unserialized graph_def
+        with tf.gfile.GFile(frozen_graph_filename, "rb") as f:
+            graph_def = tf.GraphDef()
+            graph_def.ParseFromString(f.read())
+
+        # Then, we import the graph_def into a new Graph and returns it
+        with tf.Graph().as_default() as graph:
+            # The name var will prefix every op/nodes in your graph
+            # Since we load everything in a new graph, this is not needed
+            tf.import_graph_def(graph_def, name='')
+        return graph
+
+
+
+    def init_point_cloud_array(self):
+        world_points_array = np.empty( (self.resize_height,self.resize_width) ,dtype=object)
+        for r,c in np.ndindex(self.resize_height,self.resize_width):
+            ray = np.asarray( self.camera_model.projectPixelTo3dRay( (c,r) ) )
+            ray_in_world = np.matmul( self.cam_transform_rotation_matrix, ray )
+            scale = -self.cam_transform_translation[2] / ray_in_world[2]
+            world_point = scale * ray_in_world + self.cam_transform_translation
+            world_points_array[r,c] = (world_point[0], world_point[1], world_point[2])
+        self.world_point_array = world_points_array
+
+
+    def image_callback(self, image_msg):
+        start = timer()
+        try:
+            cv_image = self.bridge.imgmsg_to_cv2(image_msg, "bgr8")
+        except CvBridgeError as e:
+            print(e)
+        cv_image = cv2.cvtColor(cv_image, cv2.COLOR_BGR2RGB)
+        cv_image = cv2.resize(cv_image, (self.resize_width, self.resize_height))
+        im = np.asarray(cv_image)
+
+        image_pl = self.graph.get_tensor_by_name('Placeholder_1:0')
+        softmax = self.graph.get_tensor_by_name('Validation/decoder/Softmax:0')
+
+        output = self.sess.run([softmax], feed_dict={image_pl: im})
+        shape = im.shape
+        output_image = output[0][:, 1].reshape(shape[0], shape[1])
+
+        threshold = 0.5
+        im_threshold = output_image > threshold
+        world_points = self.world_point_array[im_threshold].tolist()
+
+        im_threshold = np.uint8(255*im_threshold)
+        cv_output = cv2.cvtColor(im_threshold, cv2.COLOR_GRAY2BGR)
+        msg_out = self.bridge.cv2_to_imgmsg(cv_output, 'bgr8')
+        msg_out.header.stamp = image_msg.header.stamp
+        self.im_publisher.publish(msg_out)
+
+        points = world_points
+        cloud_msg = PointCloud2()
+        cloud_msg.header.stamp = image_msg.header.stamp
+        cloud_msg.header.frame_id = 'base_footprint'
+        cloud_msg.height = 1
+        cloud_msg.width = len(world_points)
+        cloud_msg.fields = [
+            PointField('x', 0, PointField.FLOAT32, 1),
+            PointField('y', 4, PointField.FLOAT32, 1),
+            PointField('z', 8, PointField.FLOAT32, 1)]
+        cloud_msg.is_bigendian = False
+        cloud_msg.point_step = 12
+        cloud_msg.row_step = 3 * len(world_points)
+        cloud_msg.data = np.asarray(world_points, np.float32).tostring()
```

Comment:
I'm not exactly sure, but is it possible to store the string array instead of list of points (when initializing earlier)? If we can, that might be slightly faster (possibly negligibly so but not sure).